crates/re_types_builder/src/codegen/cpp.rs

```diff
@@ -0,0 +1,179 @@
+            .with_context(|| format!("{folder_path:?}"))
+            .unwrap();
+
+        let mut filepaths = BTreeSet::default();
+
+        // Generate folder contents:
+        let ordered_objects = objects.ordered_objects(object_kind.into());
+        for &obj in &ordered_objects {
+            let filename = obj.snake_case_name();
+            let (hpp, cpp) = generate_hpp_cpp(objects, arrow_registry, obj);
+            for (extension, tokens) in [("hpp", hpp), ("cpp", cpp)] {
+                let string = string_from_token_stream(&tokens, obj.relative_filepath());
+                let filepath = folder_path.join(format!("{filename}.{extension}"));
+                write_file(&filepath, string);
+                let inserted = filepaths.insert(filepath);
+                assert!(
+                    inserted,
+                    "Multiple objects with the same name: {:?}",
+                    obj.name
+                );
+            }
+        }
+
+        {
+            // Generate module file that includes all the headers:
+            let hash = quote! { # };
+            let pragma_once = pragma_once();
+            let header_file_names = ordered_objects
+                .iter()
+                .map(|obj| format!("{folder_name}/{}.hpp", obj.snake_case_name()));
+            let tokens = quote! {
+                #pragma_once
+                #(#hash include #header_file_names "RE_TOKEN_NEWLINE")*
+            };
+            let filepath = folder_path
+                .parent()
+                .unwrap()
+                .join(format!("{folder_name}.hpp"));
+            let string = string_from_token_stream(&tokens, None);
+            write_file(&filepath, string);
+            filepaths.insert(filepath);
+        }
+
+        // Clean up old files:
+        for entry in std::fs::read_dir(folder_path).unwrap().flatten() {
+            let filepath = Utf8PathBuf::try_from(entry.path()).unwrap();
+            if !filepaths.contains(&filepath) {
+                std::fs::remove_file(filepath).ok();
+            }
+        }
+
+        filepaths
+    }
+}
+
+impl crate::CodeGenerator for CppCodeGenerator {
+    fn generate(
+        &mut self,
+        objects: &Objects,
+        arrow_registry: &ArrowRegistry,
+    ) -> BTreeSet<Utf8PathBuf> {
+        let mut filepaths = BTreeSet::new();
+
+        for object_kind in ObjectKind::ALL {
+            let folder_name = object_kind.plural_snake_case();
+            filepaths.extend(self.generate_folder(
+                objects,
+                arrow_registry,
+                object_kind,
+                folder_name,
+            ));
+        }
+
+        filepaths
+    }
+}
+
+fn string_from_token_stream(token_stream: &TokenStream, source_path: Option<&Utf8Path>) -> String {
+    let mut code = String::new();
+    code.push_str(&format!("// {AUTOGEN_WARNING}\n"));
+    if let Some(source_path) = source_path {
+        code.push_str(&format!("// Based on {source_path:?}"));
```

Comment:
I think we should add this to Python and Rust too. It is wonderful to be able to figure out _why_ the code was generated.